{
  "name": "StaticObjectException",
  "language": "cpp",
  "description": null,
  "example": null,
  "cwe": null,
  "cwe-description": null,
  "checker-language": "cpp",
  "loc": 55,
  "branches": 2,
  "apis": 46,
  "test": [
    {
      "description": null,
      "expected-problems": 0,
      "expected-linenumbers": [],
      "code": "\n// RUN: clang-tidy %s -checks=\"-*,cert-err58-cpp\" -- -std=c++17 -target x86_64-pc-linux-gnu \\\n// RUN:   | FileCheck %s -check-prefix=CHECK-EXCEPTIONS \\\n// RUN:   -implicit-check-not=\"{{warning|error}}:\"\n// RUN: clang-tidy %s -checks=\"-*,cert-err58-cpp\" -- -DNONEXCEPTIONS -fno-exceptions -std=c++17 -target x86_64-pc-linux-gnu \\\n// RUN:   | FileCheck %s -allow-empty -check-prefix=CHECK-NONEXCEPTIONS \\\n// RUN:   -implicit-check-not=\"{{warning|error}}:\"\n\nstruct S {\n  S() noexcept(false);\n};\n\nstruct T {\n  T() noexcept;\n};\n\nstruct U {\n  U() {}\n};\n\nstruct V {\n  explicit V(const char *) {} // Can throw\n};\n\nstruct Cleanup {\n  ~Cleanup() {}\n};\n\nstruct W {\n  W(Cleanup c = {}) noexcept(false);\n};\n\nstruct X {\n  X(S = {}) noexcept;\n};\n\nstruct Y {\n  S s;\n};\n\nstruct Z {\n  T t;\n};\n\nint f();\nint g() noexcept(false);\nint h() noexcept(true);\n\nstruct UserConv_Bad {\n  operator int() noexcept(false);\n};\n\nstruct UserConv_Good {\n  operator int() noexcept;\n};\n\nUserConv_Bad some_bad_func() noexcept;\nUserConv_Good some_good_func() noexcept;\n\nS s;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 's' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: 9:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nT t; // ok\nU u;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'u' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 17:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nV v(\"v\");\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'v' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 21:12: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nW w;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'w' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 29:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nX x1(S{});\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'x1' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 9:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nX x2;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'x2' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 9:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nY y;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'y' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 36:8: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nZ z;\n\nint i = f();\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:5: warning: initialization of 'i' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 44:5: note: possibly throwing function declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nint j = g();\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:5: warning: initialization of 'j' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 45:5: note: possibly throwing function declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nint k = h();\nint l = some_bad_func();\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:5: warning: initialization of 'l' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 49:3: note: possibly throwing function declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nint m = some_good_func();\n\ntypedef decltype(sizeof(int)) size_t;\ninline void *operator new(size_t sz, void *here) noexcept { return here; }\nchar n[sizeof(int)];\nint *o = new (n) int();\nint *p = new int();\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:6: warning: initialization of 'p' with static storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\n\nthread_local S s3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 's3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local T t3; // ok\nthread_local U u3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'u3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local V v3(\"v\");\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'v3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local W w3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'w3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\n\nvoid f(S s1, T t1, U u1, V v1, W w1) { // ok, ok, ok, ok, ok\n  S s2; // ok\n  T t2; // ok\n  U u2; // ok\n  V v2(\"v\"); // ok\n  W w2; // ok\n\n  thread_local S s3; // ok\n  thread_local T t3; // ok\n  thread_local U u3; // ok\n  thread_local V v3(\"v\"); // ok\n  thread_local W w3; // ok\n\n  static S s4; // ok\n  static T t4; // ok\n  static U u4; // ok\n  static V v4(\"v\"); // ok\n  static W w4; // ok\n}\n\nnamespace {\nS s;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 's' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: 9:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nT t; // ok\nU u;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'u' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 17:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nV v(\"v\");\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'v' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 21:12: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nW w;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:3: warning: initialization of 'w' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 29:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\n\nthread_local S s3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 's3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local T t3; // ok\nthread_local U u3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'u3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local V v3(\"v\");\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'v3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\nthread_local W w3;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:16: warning: initialization of 'w3' with thread_local storage duration may throw an exception that cannot be caught\n// CHECK-NONEXCEPTIONS-NOT: warning:\n}; // namespace\n\nclass Statics {\n  static S s; // warn when initialized\n  static T t; // ok\n  static U u; // warn when initialized\n  static V v; // warn when initialized\n  static W w; // warn when initialized\n\n  void f(S s, T t, U u, V v) {\n    S s2;      // ok\n    T t2;      // ok\n    U u2;      // ok\n    V v2(\"v\"); // ok\n    W w2;      // ok\n\n    thread_local S s3;      // ok\n    thread_local T t3;      // ok\n    thread_local U u3;      // ok\n    thread_local V v3(\"v\"); // ok\n    thread_local W w3;      // ok\n\n    static S s4;      // ok\n    static T t4;      // ok\n    static U u4;      // ok\n    static V v4(\"v\"); // ok\n    static W w4;      // ok\n  }\n};\n\nS Statics::s;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:12: warning: initialization of 's' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: 9:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nT Statics::t;\nU Statics::u;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:12: warning: initialization of 'u' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 17:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nV Statics::v(\"v\");\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:12: warning: initialization of 'v' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 21:12: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\nW Statics::w;\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:12: warning: initialization of 'w' with static storage duration may throw an exception that cannot be caught\n// CHECK-EXCEPTIONS: 29:3: note: possibly throwing constructor declared here\n// CHECK-NONEXCEPTIONS-NOT: warning:\n\n#ifndef NONEXCEPTIONS\nnamespace pr35457 {\nconstexpr int foo(int x) { if (x <= 0) throw 12; return x; }\n\nconstexpr int bar = foo(1); // OK\n// CHECK-EXCEPTIONS-NOT: warning: initialization of 'bar' with static storage\nint baz = foo(0); // Not OK; throws at runtime when exceptions are enabled.\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:5: warning: initialization of 'baz' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: :[[@LINE-6]]:15: note: possibly throwing function declared here\n} // namespace pr35457\n#endif // NONEXCEPTIONS\n\nnamespace pr39777 {\nstruct S { S(); };\nstruct T { T() noexcept; };\n\nauto Okay1 = []{ S s; };\nauto Okay2 = []{ (void)new int; };\nauto NotOkay1 = []{ S s; return 12; }(); // Because the lambda call is not noexcept\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:6: warning: initialization of 'NotOkay1' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: :[[@LINE-7]]:12: note: possibly throwing constructor declared here\nauto NotOkay2 = []() noexcept { S s; return 12; }(); // Because S::S() is not noexcept\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:6: warning: initialization of 'NotOkay2' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: :[[@LINE-10]]:12: note: possibly throwing constructor declared here\nauto Okay3 = []() noexcept { T t; return t; }();\n\nstruct U {\n  U() noexcept;\n  auto getBadLambda() const noexcept {\n    return []{ S s; return s; };\n  }\n};\nauto Okay4 = []{ U u; return u.getBadLambda(); }();\nauto NotOkay3 = []() noexcept { U u; return u.getBadLambda(); }()(); // Because the lambda returned and called is not noexcept\n// CHECK-EXCEPTIONS: :[[@LINE-1]]:6: warning: initialization of 'NotOkay3' with static storage duration may throw an exception that cannot be caught [cert-err58-cpp]\n// CHECK-EXCEPTIONS: :[[@LINE-6]]:12: note: possibly throwing function declared here\n\n#ifndef NONEXCEPTIONS\nstruct Bad {\n  Bad() {\n    throw 12;\n  }\n};\n\nstatic auto NotOkay4 = [bad = Bad{}](){};\n// FIXME: the above should be diagnosed because the capture init can trigger\n// an exception when constructing the Bad object.\n#endif // NONEXCEPTIONS\n}"
    }
  ]
}